texlive.withPackages: generate formats in separate derivations

xworld21 opened the pull request:
## Description of changes
Generate formats in separate derivations, so that they do not get rebuild for each `texlive.withPackages` call. It follows the same logic set by tlmgr: a format depends only on its dependencies and the hyphenation patterns in the system (and the paper size, but that's not configurable yet).

This should make quite a difference, since recompiling the formats can take more than a minute with the large schemes.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

veprbl commented:
@GrahamcOfBorg build tests.texlive

veprbl reviewed (approved):
nixpkgs-review looks fine

xworld21 committed:
texlive.withPackages: generate formats in separate derivations

Generate TeX formats into separate derivations so that they can be
shared across different environments, considerably reducing build times.

veprbl commented:
Result of `nixpkgs-review pr 291396` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages marked as broken and skipped:</summary>
  <ul>
    <li>dwarf-fortress-packages.dwarf-therapist-original</li>
    <li>golden-cheetah-bin</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_5.rdc</li>
    <li>rocmPackages_5.rdc.doc</li>
    <li>rocmPackages_6.mivisionx</li>
    <li>rocmPackages_6.mivisionx-cpu</li>
    <li>rocmPackages_6.mivisionx-hip</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
    <li>spring</li>
    <li>springLobby</li>
  </ul>
</details>
<details>
  <summary>22 packages failed to build:</summary>
  <ul>
    <li>python311Packages.abjad</li>
    <li>python311Packages.abjad.dist</li>
    <li>python311Packages.torchWithRocm</li>
    <li>python311Packages.torchWithRocm.cxxdev</li>
    <li>python311Packages.torchWithRocm.dev</li>
    <li>python311Packages.torchWithRocm.dist</li>
    <li>python311Packages.torchWithRocm.lib</li>
    <li>python312Packages.abjad</li>
    <li>python312Packages.abjad.dist</li>
    <li>python312Packages.pplpy</li>
    <li>python312Packages.pplpy.dist</li>
    <li>python312Packages.pplpy.doc</li>
    <li>python312Packages.torchWithRocm</li>
    <li>python312Packages.torchWithRocm.cxxdev</li>
    <li>python312Packages.torchWithRocm.dev</li>
    <li>python312Packages.torchWithRocm.dist</li>
    <li>python312Packages.torchWithRocm.lib</li>
    <li>rocmPackages_5.mivisionx (rocmPackages_5.mivisionx-hip)</li>
    <li>rocmPackages_5.mivisionx-cpu</li>
    <li>rocmPackages_5.mivisionx-opencl</li>
    <li>rocmPackages_5.rocprofiler</li>
    <li>sageWithDoc</li>
  </ul>
</details>
<details>
  <summary>243 packages built:</summary>
  <ul>
    <li>R</li>
    <li>R.tex</li>
    <li>advi</li>
    <li>apostrophe</li>
    <li>asciidoc-full</li>
    <li>asciidoc-full-with-plugins</li>
    <li>asciidoc-full-with-plugins.dist</li>
    <li>asciidoc-full.dist</li>
    <li>asymptote</li>
    <li>asymptote.doc</li>
    <li>asymptote.info</li>
    <li>asymptote.man</li>
    <li>asymptote.tex</li>
    <li>auctex</li>
    <li>auctex.tex</li>
    <li>auto-multiple-choice</li>
    <li>avrdude</li>
    <li>avrdudess</li>
    <li>blahtexml</li>
    <li>blahtexml.doc</li>
    <li>bluespec</li>
    <li>bluespec.doc</li>
    <li>cantor (libsForQt5.cantor ,plasma5Packages.cantor)</li>
    <li>catdvi</li>
    <li>catdvi.dev</li>
    <li>cddlib</li>
    <li>clevis</li>
    <li>clevis.man</li>
    <li>dblatex</li>
    <li>dblatexFull</li>
    <li>denemo</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>digikam</li>
    <li>disorderfs</li>
    <li>dot2tex (python311Packages.dot2tex)</li>
    <li>dot2tex.dist (python311Packages.dot2tex.dist)</li>
    <li>dwarf-fortress-packages.dwarf-fortress-full</li>
    <li>dwarf-therapist (dwarf-fortress-packages.dwarf-therapist)</li>
    <li>eclib</li>
    <li>enblend-enfuse</li>
    <li>extractpdfmark</li>
    <li>fastcap</li>
    <li>fastcap.doc</li>
    <li>frescobaldi</li>
    <li>frescobaldi.dist</li>
    <li>gap</li>
    <li>gap-full</li>
    <li>gap-minimal</li>
    <li>gfan</li>
    <li>giac</li>
    <li>giac-with-xcas</li>
    <li>giac.doc</li>
    <li>gnu-cobol</li>
    <li>gnu-cobol.bin</li>
    <li>gnu-cobol.dev</li>
    <li>gnu-cobol.lib</li>
    <li>gp2c</li>
    <li>herwig</li>
    <li>hid-tools</li>
    <li>hid-tools.dist</li>
    <li>hugin</li>
    <li>inkscape-extensions.textext</li>
    <li>inkscape-extensions.textext.dist</li>
    <li>iotas</li>
    <li>ipe</li>
    <li>jasp-desktop</li>
    <li>jetbrains.dataspell</li>
    <li>rep (kakounePlugins.rep)</li>
    <li>klipper-firmware</li>
    <li>klipper-flash</li>
    <li>labplot</li>
    <li>latte-integrale</li>
    <li>lcalc</li>
    <li>lilypond</li>
    <li>lilypond-unstable</li>
    <li>lilypond-unstable-with-fonts</li>
    <li>lilypond-with-fonts</li>
    <li>luksmeta</li>
    <li>manim</li>
    <li>manim.dist</li>
    <li>mitscheme</li>
    <li>mitschemeX11</li>
    <li>ne</li>
    <li>nuweb</li>
    <li>openmolcas</li>
    <li>opentx</li>
    <li>paperwork</li>
    <li>paperwork.dist</li>
    <li>pari</li>
    <li>polymake</li>
    <li>postgresql12JitPackages.plr</li>
    <li>postgresql12Packages.plr</li>
    <li>postgresql13JitPackages.plr</li>
    <li>postgresql13Packages.plr</li>
    <li>postgresql14JitPackages.plr</li>
    <li>postgresql14Packages.plr</li>
    <li>postgresqlJitPackages.plr (postgresql15JitPackages.plr)</li>
    <li>postgresqlPackages.plr (postgresql15Packages.plr)</li>
    <li>postgresql16JitPackages.plr</li>
    <li>postgresql16Packages.plr</li>
    <li>pynac</li>
    <li>python311Packages.cnvkit</li>
    <li>python311Packages.cnvkit.dist</li>
    <li>python311Packages.cypari2</li>
    <li>python311Packages.cypari2.dist</li>
    <li>python311Packages.cysignals</li>
    <li>python311Packages.cysignals.dist</li>
    <li>python311Packages.fpylll</li>
    <li>python311Packages.fpylll.dist</li>
    <li>python311Packages.nbdev</li>
    <li>python311Packages.nbdev.dist</li>
    <li>python311Packages.pplpy</li>
    <li>python311Packages.pplpy.dist</li>
    <li>python311Packages.pplpy.doc</li>
    <li>python311Packages.primecountpy</li>
    <li>python311Packages.primecountpy.dist</li>
    <li>python311Packages.pylatex</li>
    <li>python311Packages.pylatex.dist</li>
    <li>python311Packages.pypandoc</li>
    <li>python311Packages.pypandoc.dist</li>
    <li>python311Packages.pytikz-allefeld</li>
    <li>python311Packages.pytikz-allefeld.dist</li>
    <li>python311Packages.radian</li>
    <li>python311Packages.radian.dist</li>
    <li>python311Packages.rchitect</li>
    <li>python311Packages.rchitect.dist</li>
    <li>recoll (python311Packages.recoll)</li>
    <li>rivet (python311Packages.rivet)</li>
    <li>python311Packages.rpy2</li>
    <li>python311Packages.rpy2.dist</li>
    <li>python311Packages.sphinxcontrib-bayesnet</li>
    <li>python311Packages.sphinxcontrib-bayesnet.dist</li>
    <li>python311Packages.sphinxcontrib-tikz</li>
    <li>python311Packages.sphinxcontrib-tikz.dist</li>
    <li>python311Packages.unstructured</li>
    <li>python311Packages.unstructured.dist</li>
    <li>python312Packages.cypari2</li>
    <li>python312Packages.cypari2.dist</li>
    <li>python312Packages.cysignals</li>
    <li>python312Packages.cysignals.dist</li>
    <li>python312Packages.dot2tex</li>
    <li>python312Packages.dot2tex.dist</li>
    <li>python312Packages.fpylll</li>
    <li>python312Packages.fpylll.dist</li>
    <li>python312Packages.nbdev</li>
    <li>python312Packages.nbdev.dist</li>
    <li>python312Packages.primecountpy</li>
    <li>python312Packages.primecountpy.dist</li>
    <li>python312Packages.pylatex</li>
    <li>python312Packages.pylatex.dist</li>
    <li>python312Packages.pypandoc</li>
    <li>python312Packages.pypandoc.dist</li>
    <li>python312Packages.pytikz-allefeld</li>
    <li>python312Packages.pytikz-allefeld.dist</li>
    <li>python312Packages.radian</li>
    <li>python312Packages.radian.dist</li>
    <li>python312Packages.rchitect</li>
    <li>python312Packages.rchitect.dist</li>
    <li>python312Packages.recoll</li>
    <li>python312Packages.rivet</li>
    <li>python312Packages.rpy2</li>
    <li>python312Packages.rpy2.dist</li>
    <li>python312Packages.sphinxcontrib-bayesnet</li>
    <li>python312Packages.sphinxcontrib-bayesnet.dist</li>
    <li>python312Packages.sphinxcontrib-tikz</li>
    <li>python312Packages.sphinxcontrib-tikz.dist</li>
    <li>python312Packages.unstructured</li>
    <li>python312Packages.unstructured.dist</li>
    <li>qmk</li>
    <li>qmk.dist</li>
    <li>quarto</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages_5.migraphx</li>
    <li>rocmPackages_5.miopen (rocmPackages_5.miopen-hip)</li>
    <li>rocmPackages_5.miopen-opencl</li>
    <li>rocmPackages_5.miopengemm</li>
    <li>rocmPackages_5.miopengemm.doc</li>
    <li>rocmPackages_5.rocdbgapi</li>
    <li>rocmPackages_5.rocdbgapi.doc</li>
    <li>rocmPackages_5.rocr-debug-agent</li>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
    <li>sage</li>
    <li>sagetex.tex</li>
    <li>sherpa</li>
    <li>singular</li>
    <li>skribilo</li>
    <li>solfege</li>
    <li>spark (spark3 ,spark_3_5)</li>
    <li>spark_3_4</li>
    <li>sympow</li>
    <li>tang</li>
    <li>tang.man</li>
    <li>tests.pkg-config.defaultPkgConfigPackages.libR</li>
    <li>texliveBasic</li>
    <li>texliveBasic.info</li>
    <li>texliveBasic.man</li>
    <li>texliveBookPub</li>
    <li>texliveBookPub.info</li>
    <li>texliveBookPub.man</li>
    <li>texliveConTeXt</li>
    <li>texliveConTeXt.info</li>
    <li>texliveConTeXt.man</li>
    <li>texliveFull</li>
    <li>texliveFull.info</li>
    <li>texliveFull.man</li>
    <li>texliveGUST</li>
    <li>texliveGUST.info</li>
    <li>texliveGUST.man</li>
    <li>texliveInfraOnly</li>
    <li>texliveInfraOnly.info</li>
    <li>texliveInfraOnly.man</li>
    <li>texliveMedium</li>
    <li>texliveMedium.info</li>
    <li>texliveMedium.man</li>
    <li>texliveMinimal</li>
    <li>texliveMinimal.info</li>
    <li>texliveMinimal.man</li>
    <li>texliveSmall</li>
    <li>texliveSmall.info</li>
    <li>texliveSmall.man</li>
    <li>texliveTeTeX</li>
    <li>texliveTeTeX.info</li>
    <li>texliveTeTeX.man</li>
    <li>texmacs</li>
    <li>thepeg</li>
    <li>therion</li>
    <li>tinygo</li>
    <li>wayback-machine-archiver</li>
    <li>wayback-machine-archiver.dist</li>
    <li>xyce</li>
    <li>xyce-parallel</li>
    <li>xyce-parallel.doc</li>
    <li>xyce.doc</li>
    <li>yosys-bluespec</li>
    <li>zettlr</li>
  </ul>
</details>
